### PR TITLE
Update rails.md

### PR DESCRIPTION
**Pull request by leymannx**
### Proposed changes

It occupied me quite a while figuring out why this sample didn't have a restart policy. And why we where supposed to always properly stop the containers wheres you normally would simply hit `Ctrl+C`.

Turned out that this `pid` issue is a rails-specific issue. https://github.com/docker/compose/issues/1393#issuecomment-99988242

And of course there's already [a Gem for fixing that](https://github.com/lorenzosinisi/shutup) but having the one-liner as entry command in `docker-compose.yml` keeps the logic where it belongs to (and now also is treated here in the sample).

### Related issues (optional)

https://github.com/docker/compose/issues/1393#issuecomment-99988242

### CODE REVIEW

**Comment by leymannx** [comment]
🤔 Or maybe we should skip `restart: always`? As this is out of the scope of this tutorial.

**Comment by leymannx** [comment]
Removed `restart: always` again.

**Comment by thaJeztah** [comment]
Thanks! I think this page could use some additional changes; perhaps you're interested in working on that.


First of all (related to this change); To stick to "best practices", I think the image "consumer" should not have to deal with this, and it's the responsibility of the image _author_ to make sure the image functions correctly (and take cleaning up old .pid files into account).

Given that this page also shows how to build the image, I think we should add an entrypoint script to the example that does this cleanup step (also see my comment https://github.com/docker/compose/issues/1393#issuecomment-100093609))

The entrypoint script could look like;

```bash
#!/bin/bash
set -e

# Rails gets grumpy about PID files pre-existing
rm -f /myapp/tmp/pids/server.pid

exec "$@"
```

That script would automatically cleanup the `.pid` file when the container is started, and after cleaning up, `exec` the container's main process (what's set as `CMD`), if set as `ENTRYPOINT`. Something like:

```Dockerfile
COPY entrypoint.sh /usr/bin/
RUN chmod +x /usr/bin/entrypoint.sh
ENTRYPOINT ["entrypoint.sh"]
```


Looking at the Dockerfile, I notice that the version of ruby is pretty outdated now. It also doesn't have a new `CMD` set to actually run `rails` (instead of the `CMD` of the `ruby` image). Because of that, the `docker-compose.yml` has to set the correct command. This should not be needed; if this image is built to run rails, it should do so automatically :smile:

Now, bear with me, because I'm not a Ruby / Rails developer, but perhaps we should update the Dockerfile to match [the example Dockerfile from the official image](https://github.com/docker-library/rails);

```Dockerfile
FROM ruby:2.5

RUN apt-get update \
        && apt-get install -y --no-install-recommends \
                nodejs \
                postgresql-client \
        && rm -rf /var/lib/apt/lists/*

COPY entrypoint.sh /usr/bin/
RUN chmod +x /usr/bin/entrypoint.sh
WORKDIR /myapp
COPY Gemfile* ./
RUN bundle install
COPY . .

EXPOSE 3000
ENTRYPOINT ["entrypoint.sh"]

# Listen on 0.0.0.0 so that it's accessible outside the container
CMD ["rails", "server", "-b", "0.0.0.0"]
```

While we're at it, probably update the version of rails in the inital Gemfile as well;

```
source 'https://rubygems.org'
gem 'rails', '5.1.5.0'
```

Now that the Dockerfile has a default command set, it's no longer needed to override the command in the `docker-compose.yml`, so the whole `command:` line can be removed:

```
command: bash -c "rm -f tmp/pids/server.pid && bundle exec rails s -p 3000 -b '0.0.0.0'"
```

**Comment by leymannx** [comment]
Wow thanks for the insights! I'll doctor around a little bit to get everything working and then come back to incorporate the updates.

**Comment by leymannx** [comment]
Works super smooth when done as you proposed. Only thing that makes me wonder is that after setting everything up and hitting `docker-compose up` it only shows the `db` container:

```
Starting rails_db_1 ... done
Starting rails_web_1 ... done
Attaching to rails_db_1, rails_web_1
db_1   | 2018-03-21 19:46:12.240 UTC [1] LOG:  listening on IPv4 address "0.0.0.0", port 5432
db_1   | 2018-03-21 19:46:12.240 UTC [1] LOG:  listening on IPv6 address "::", port 5432
db_1   | 2018-03-21 19:46:12.245 UTC [1] LOG:  listening on Unix socket "/var/run/postgresql/.s.PGSQL.5432"
db_1   | 2018-03-21 19:46:12.384 UTC [22] LOG:  database system was shut down at 2018-03-21 19:45:43 UTC
db_1   | 2018-03-21 19:46:12.428 UTC [1] LOG:  database system is ready to accept connections
```

And then only after I accessed http://localhost:3000 in the browser the `web` container kicks in:

```
web_1  | => Booting Puma
web_1  | => Rails 5.1.5 application starting in development 
web_1  | => Run `rails server -h` for more startup options
web_1  | Puma starting in single mode...
web_1  | * Version 3.11.3 (ruby 2.5.0-p0), codename: Love Song
web_1  | * Min threads: 5, max threads: 5
web_1  | * Environment: development
web_1  | * Listening on tcp://0.0.0.0:3000
web_1  | Use Ctrl-C to stop
web_1  | Started GET "/" for 172.20.0.1 at 2018-03-21 19:47:21 +0000
web_1  | Cannot render console from 172.20.0.1! Allowed networks: 127.0.0.1, ::1, 127.0.0.0/127.255.255.255
web_1  | Processing by Rails::WelcomeController#index as HTML
web_1  |   Rendering /usr/local/bundle/gems/railties-5.1.5/lib/rails/templates/rails/welcome/index.html.erb
web_1  |   Rendered /usr/local/bundle/gems/railties-5.1.5/lib/rails/templates/rails/welcome/index.html.erb (8.2ms)
web_1  | Completed 200 OK in 918ms (Views: 26.7ms)
web_1  | 
web_1  | 
```

This is different from the old example. So I stripped that out for now.

Do you think that's important?

**Comment by ahh-docker** [comment]
@shin- PTAL.

File: compose/rails.md [hunk 1]
```diff
@@ -151,35 +177,28 @@ You can now boot the app with [docker-compose up](/compose/reference/up/):
...
+```
 
 Finally, you need to create the database. In another terminal, run:
 
     docker-compose run web rake db:create
 
 Here is an example of the output from that command:
 
-```none
+```shell
```
Comment (shin-) [inline compose/rails.md]: I think the `shell` syntax hint denotes a shell script and isn't appropriate for shell output. [GFM lets you use `console`](https://stackoverflow.com/a/49004070), but I don't know if that works with the engine we use to render our docs.

File: compose/rails.md [hunk 2]
```diff
@@ -64,10 +74,25 @@ to link them together and expose the web app's port.
...
+exec "$@"
+```
 
 ### Build the project
 
-With those four files in place, you can now generate the Rails skeleton app
+With those five files in place, you can now generate the Rails skeleton app
 using [docker-compose run](/compose/reference/run/):
 
     docker-compose run web rails new . --force --database=postgresql
```
[the remark this answers is not in the document]
Comment (paigehargrave) [reply compose/rails.md]: @leymannx - Hi Norman, I don't have access to edit the file - can you address Joffrey's comments? Thank you!

[the remark this answers is not in the document]
Comment (thaJeztah) [reply compose/rails.md]: @paigehargrave looks like the fork was removed, so we'll have to restore the branch

ecbb4c4f-9139-45d5-939c-17c800f64459

[the remark this answers is not in the document]
Comment (bermudezmt) [reply compose/rails.md]: Suggestion added in #7965

File: compose/rails.md (hunk 1 above)
Comment (bermudezmt) [reply compose/rails.md]: Fixed in #7965

File: compose/rails.md [hunk 3]
```diff
@@ -9,21 +9,30 @@ a Rails/PostgreSQL app. Before starting, [install Compose](install.md).
 
 ### Define the project
 
-Start by setting up the four files needed to build the app. First, since
+Start by setting up the five files needed to build the app. First, since
```
[the remark this answers is not in the document]
Comment (bermudezmt) [reply compose/rails.md]: Fixed in #7965

File: compose/rails.md [hunk 4]
```diff
@@ -64,10 +74,25 @@ to link them together and expose the web app's port.
 
 >**Tip**: You can use either a `.yml` or `.yaml` extension for this file.
 
+Finally, we'll provide an entrypoint script to fix a Rails-specific issue that
```
[the remark this answers is not in the document]
Comment (bermudezmt) [reply compose/rails.md]: Incorporated in #7966

**Comment by bermudezmt** [comment]
Fixed gnarly merge conflicts and incorporated feedback via #7965 and #7966 . Thank you, @leymannx , @thaJeztah , @shin- , and @paigehargrave for all the work put into this!

**Pull request closed without merging**